### PR TITLE
[docs] add READMEs of supported solvers to the documentation

### PR DESCRIPTION
**Pull request by odow**
Closes https://github.com/jump-dev/JuMP.jl/issues/3307.

I'll add a preview link once CI finishes: https://jump.dev/JuMP.jl/previews/PR3309/packages/AmplNLWriter/

There's a few critical blockers, most noticeably, the Markdown supported by GitHub READMEs is different to Documenter's Markdown...

Compare

image

to

image

So perhaps _we_ need to control the documentation within JuMP? Thinking about it though, having all of the solvers within the JuMP documentation would make a big difference to new users. (I've been using Pyomo for a different project recently, and I struggled with just about everything related to using and installing a solver.)

### CODE REVIEW

**Comment by odow** [comment]
I guess we can control all of the READMEs within `jump-dev`, and we just say that any external solvers need to have a compatible README, or they need to provide a separate `JuMP.md` file that we can pull in if they want to appear in the docs.

**Comment by odow** [comment]
The other option is to use https://github.com/JuliaComputing/MultiDocumenter.jl like the SciML docs; https://docs.sciml.ai/Overview/stable/, but that seems overkill. It's okay if the solvers have their own documentation and the README is just a link to that.

**Comment by odow** [comment]
```Julia
import MathOptInterface as MOI

function print_supported(optimizer; eltype = Float64)
    model = MOI.instantiate(optimizer)
    functions, sets, model_attrs = Any[], Any[], Any[]
    for name in names(MOI; all = true)
        x = getfield(MOI, name)
        if !(x isa Type)
            continue
        end
        if !isconcretetype(x)
            try
                x = x{eltype}
            catch
                continue
            end    
        end
        if x <: MOI.AbstractSet
            push!(sets, x)
        elseif x <: MOI.AbstractFunction
            push!(functions, x)
        elseif x <: MOI.AbstractModelAttribute
            push!(model_attrs, x)
        end
    end
    for F in functions
        push!(model_attrs, MOI.ObjectiveFunction{F})
    end
    supported_constraints = Any[
        replace(" * [`$F`](@ref) in [`$S`](@ref)", "MathOptInterface" => "MOI")
        for S in sets, F in functions if MOI.supports_constraint(model, F, S)
    ]
    sort!(supported_constraints)
    supported_model_attrs = Any[
        replace(" * [`$attr`](@ref)", "MathOptInterface" => "MOI")
        for attr in model_attrs if (try MOI.supports(model, attr()); catch; false end)
    ]
    sort!(supported_model_attrs)
    println(
        "List of supported model attributes:\n\n",
        join(supported_model_attrs, "\n"),
        "\n\nList of supported constraint types:\n\n",
        join(supported_constraints, "\n"),
    )
end

import AmplNLWriter, CPLEX, Cbc, Clp
print_supported(AmplNLWriter.Optimizer)
print_supported(CPLEX.Optimizer)
print_supported(Cbc.Optimizer)
print_supported(Clp.Optimizer)
```

File: docs/solvers.toml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+#   user = "oxfordcontrol"
+[Clp]
+    rev = "od/readme"
+# [COPT]
+#   user = "COPT-Public"
+# [COSMO]
+#   user = "oxfordcontrol"
+[CPLEX]
+    rev = "od/readme"
+[CSDP]
```
Comment (odow) [inline docs/solvers.toml:23]:
```suggestion
[CSDP]
    rev = "od/readme"
```

File: docs/solvers.toml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+# [DAQP]
+#   user = "darnstrom"
+#   rev = "main"
+[DiffOpt]
+[Dualization]
+# [EAGO]
+#   user = "PSORLab"
+[ECOS]
```
Comment (odow) [inline docs/solvers.toml:31]:
```suggestion
[ECOS]
    rev = "od/readme"
```

File: docs/solvers.toml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+# [GAMS]
+#   user = "GAMS-dev"
+[GLPK]
```
Comment (odow) [inline docs/solvers.toml:34]:
```suggestion
[GLPK]
    rev = "od/readme"
```

File: docs/solvers.toml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+[Gurobi]
```
Comment (odow) [inline docs/solvers.toml:35]:
```suggestion
[Gurobi]
    rev = "od/readme"
```

File: docs/solvers.toml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+# [COSMO]
+#   user = "oxfordcontrol"
+[CPLEX]
+    rev = "od/readme"
+[CSDP]
+    rev = "od/readme"
+# [DAQP]
+#   user = "darnstrom"
+#   rev = "main"
+[DiffOpt]
```
Comment (odow) [inline docs/solvers.toml:28]:
```suggestion
# [DiffOpt]
```

File: docs/solvers.toml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+[Dualization]
```
Comment (odow) [inline docs/solvers.toml:29]:
```suggestion
# [Dualization]
```

File: docs/solvers.toml [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+#   user = "PSORLab"
+[ECOS]
+    rev = "od/readme"
+# [GAMS]
+#   user = "GAMS-dev"
+[GLPK]
+    rev = "od/readme"
+[Gurobi]
+    rev = "od/readme"
+[HiGHS]
```
Comment (odow) [inline docs/solvers.toml:40]:
```suggestion
[HiGHS]
    rev = "od/readme"
```

File: docs/solvers.toml [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+# [Hypatia]
+#   user = "chriscoey"
+[Ipopt]
```
Comment (odow) [inline docs/solvers.toml:43]:
```suggestion
[Ipopt]
    rev = "od/readme"
```

File: docs/solvers.toml [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+# [Juniper]
+#   user = "lanl-ansi"
+[KNITRO]
+# [MadNLP]
+#   user = "MadNLP"
+[MiniZinc]
```
Comment (odow) [inline docs/solvers.toml:49]:
```suggestion
[MiniZinc]
    rev = "od/readme"
```

File: docs/solvers.toml [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    rev = "od/readme"
+[MosekTools]
+[MultiObjectiveAlgorithms]
+# [NLopt]
+#   user = "JuliaOpt"
+# [OSQP]
+#   user = "osqp"
+# [PATHSolver]
+#   user = "chkwon"
+[Pajarito]
```
Comment (odow) [inline docs/solvers.toml:61]:
```suggestion
[Pajarito]
    rev = "main"
```

File: docs/solvers.toml [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+[Pavito]
+[Penopt]
+# [ProxSDP]
+#   user = "mariohsouto"
+# [SCIP]
+#   user = "scipopt"
+[SCS]
```
Comment (odow) [inline docs/solvers.toml:68]:
```suggestion
[SCS]
    rev = "od/readme"
```

**Comment by odow** [comment]
So I think this is the right general direction. We don't want the READMEs changing on an arbitrary JuMP commit, so I'll lock them all to a specific commit once we have them in a nice place. And then periodically we can bump the commits and rebuild, just like I do with the MOI docs.

**Comment by blegat** [comment]
I like the added list of supported things in the READMEs

**Comment by odow** [comment]
> I like the added list of supported things in the READMEs

Yeah. Although is it too complicated that it's all MOI stuff instead of JuMP? And it also doesn't cover sets supported by bridging. I tried that, but it got out of hand.

**Comment by blegat** [comment]
> Although is it too complicated that it's all MOI stuff instead of JuMP?
> And it also doesn't cover sets supported by bridging. I tried that, but it got out of hand.

Yes, that would be too large. We already have the class of problems solved in the solver table. This is for people wanting more advanced details, e.g., if they are wondering whether they should use `Dualization`, playing with `print_active_bridges`, etc...

**Comment by odow** [comment]
Dear all,

I'm in the process of adding solver READMEs to the JuMP documentation:

https://jump.dev/JuMP.jl/previews/PR3309/solvers/AmplNLWriter/

I think this will make solvers more discoverable and easier to install for new users (who might not be comfortable browsing GitHub), and it will push us to write more comprehensive, uniform documentation for each wrapper (for example, making it clear what license each package is released under, providing options, ...).

The way this works is that the `docs/make.jl` script pulls in the README.md file from each solver repository and adds that to the HTML Documenter build (but not the LaTeX PDF). Before merging this PR, I'll change to pulling in a specific commit, so that changes in the solver repositories don't break the JuMP documentation. We'll periodically update the commits to update the documentation.

The packages in `jump-dev` are fine, because I can easily control what happens there. it's not so obvious what to do for third-party packages.

The main stumbling block is that the syntax supported by GitHub markdown is not the same as the syntax supported by Documenter Markdown (automatic detection of embedded HTML is one example). It also breaks local links in the README that point to files in your repositories.

I see a few options, but I'd like peoples feedback on what we should do:

 1. Force third-parties to conform their READMEs to the JuMP style if they want to be included in the documentation
 2. Allow third-parties to provide a `/JuMP_README.md` file that we pull in instead. You control what content this has.
 3. Don't include third-party solver wrappers in the JuMP documentation
 4. Provide a single page of third party solvers, which each solver is just a link to the GitHub repository (e.g., essentially just the supported solvers table).

Thoughts?

cc'ing impacted people: 

@ccoffrin @goulart-paul @lschork2 @darnstrom @renkekuhlmann @chriscoey @Wikunia @sshin23 @mariohsouto @matbesancon @mtanneau

**Comment by Wikunia** [comment]
Thanks for the ping. I would vote for option 2 to be able to have a dedicated short description for a solver inside the JuMP documentation with still having the ability for longer general readmes. Though option 4 sounds also fine to me.  
As always thanks a lot for your work @odow 😉

**Comment by renkekuhlmann** [comment]
Thanks @odow! I think option 2 works best.

**Comment by sshin23** [comment]
Thanks, @odow! Either 2 or 4 sounds good to me

File: docs/solvers.toml [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+#   user = "GAMS-dev"
+# [Hypatia]
+#   user = "chriscoey"
+# [Juniper]
+#   user = "lanl-ansi"
+# [MadNLP]
+#   user = "MadNLP"
+[NEOSServer]
+  user = "odow"
+  rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:88]:
```suggestion
  rev = "a9fdb3f"
```

**Comment by odow** [comment]
Seems pretty clear for option 2. There's a `solver.toml` file. Each package must have the format:
```
[PATHSolver]
   user = "chkwon"
   rev = "v1.0.0"
   filename = "README.md"
```
and then this will pull in the file at `https://raw.githubusercontent.com/chkwon/PATHSolver.jl/v1.0.0/README.md`

Change `rev` and `filename` as needed.

File: docs/solvers.toml [hunk 13]
```diff
@@ -0,0 +1,103 @@
+# ============================= jump-dev packages ============================ #
+
+[AmplNLWriter]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:4]:
```suggestion
    rev = "v1.1.0"
```

File: docs/solvers.toml [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+[BARON]
+    rev = "od/readme"
+[Cbc]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:8]:
```suggestion
    rev = "a78e129"
```

File: docs/solvers.toml [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+[Clp]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:10]:
```suggestion
    rev = "3358ee6"
```

File: docs/solvers.toml [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+[CPLEX]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:12]:
```suggestion
    rev = "8e6f211"
```

File: docs/solvers.toml [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+[CSDP]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:14]:
```suggestion
    rev = "3a2e4c7"
```

File: docs/solvers.toml [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+[DiffOpt]
+    rev = "od/readme"
+[Dualization]
+    rev = "od/readme"
+[ECOS]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:20]:
```suggestion
    rev = "80d1a927b05b04360b7a8e6800f5ae13bc6a1494"
```

File: docs/solvers.toml [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+[GLPK]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:22]:
```suggestion
    rev = "dd835dc425fe8995d965da8a57b4a3a1764cf8da"
```

File: docs/solvers.toml [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+[Pajarito]
+    rev = "od/readme"
+[Pavito]
+    rev = "od/readme"
+[Penopt]
+    rev = "od/readme"
+[SCS]
+    rev = "od/readme"
+[SDPA]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:47]:
```suggestion
    rev = "5d15d3c292cd25b461f524aaaf8eb5ecb8e08b0d"
```

File: docs/solvers.toml [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+[SDPNAL]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:49]:
```suggestion
    rev = "87b516fd3db316de8aa12681295138df335feed4"
```

File: docs/solvers.toml [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+[Ipopt]
+    rev = "od/readme"
+[KNITRO]
+    rev = "od/readme"
+[MiniZinc]
+    rev = "od/readme"
+[MosekTools]
+    rev = "od/readme"
+[MultiObjectiveAlgorithms]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:36]:
```suggestion
    rev = "b4413a1ef3b7d28045cedba4e68e5e4959c8bb12"
```

File: docs/solvers.toml [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+[GLPK]
+    rev = "od/readme"
+[Gurobi]
+    rev = "od/readme"
+[HiGHS]
+    rev = "od/readme"
+[Ipopt]
+    rev = "od/readme"
+[KNITRO]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:30]:
```suggestion
    rev = "3fb7c61e886b1b06d9f5cd646e194da638c93355"
```

File: docs/solvers.toml [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+[Dualization]
+    rev = "od/readme"
+[ECOS]
+    rev = "od/readme"
+[GLPK]
+    rev = "od/readme"
+[Gurobi]
+    rev = "od/readme"
+[HiGHS]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:26]:
```suggestion
    rev = "b0ab8361558fc63873a5cf380d335409f02d8863"
```

File: docs/solvers.toml [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+[Penopt]
+    rev = "od/readme"
+[SCS]
+    rev = "od/readme"
+[SDPA]
+    rev = "5d15d3c292cd25b461f524aaaf8eb5ecb8e08b0d"
+[SDPNAL]
+    rev = "87b516fd3db316de8aa12681295138df335feed4"
+[SDPT3]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:51]:
```suggestion
    rev = "02f69fb87941691722176f77a48933ff7f7a78b4"
```

File: docs/solvers.toml [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+  user = "odow"
+  rev = "v1.0.2"
+  filename = "README.md"
+# [NLopt]
+#   user = "JuliaOpt"
+# [OSQP]
+#   user = "osqp"
+[PATHSolver]
+  user = "chkwon"
+  rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:96]:
```suggestion
  rev = "08848c165747695e7a4dcbb86a657e2347ca7149"
```

File: docs/make.jl [hunk 27]
```diff
@@ -102,6 +104,55 @@ if !_FAST
...
+const _SOLVER_DIR = joinpath(@__DIR__, "src", "solvers")
+if isdir(_SOLVER_DIR)
+    rm(_SOLVER_DIR; recursive = true)
+end
+mkdir(_SOLVER_DIR)
+const _LIST_OF_SOLVERS = Pair{String,String}[]
+for (solver, data) in TOML.parsefile(joinpath(@__DIR__, "solvers.toml"))
+    user = get(data, "user", "jump-dev")
+    repo = "$solver.jl"
+    tag = get(data, "rev", "master")
```
Comment (odow) [inline docs/make.jl:120]:
```suggestion
    tag = data["rev"]
```

File: docs/solvers.toml [hunk 28]
```diff
@@ -0,0 +1,103 @@
+# ============================= jump-dev packages ============================ #
+
+[AmplNLWriter]
+    rev = "od/readme"
+[BARON]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:6]:
```suggestion
    rev = "f5f4c9c8f27e0ef8644d65fb956f3bca5305b45e"
```

File: docs/solvers.toml [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+[ECOS]
+    rev = "80d1a927b05b04360b7a8e6800f5ae13bc6a1494"
+[GLPK]
+    rev = "dd835dc425fe8995d965da8a57b4a3a1764cf8da"
+[Gurobi]
+    rev = "od/readme"
+[HiGHS]
+    rev = "b0ab8361558fc63873a5cf380d335409f02d8863"
+[Ipopt]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:28]:
```suggestion
    rev = "v1.2.1"
```

File: docs/solvers.toml [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+    rev = "b4413a1ef3b7d28045cedba4e68e5e4959c8bb12"
+    has_html = true
+[Pajarito]
+    rev = "od/readme"
+[Pavito]
+    rev = "od/readme"
+[Penopt]
+    rev = "od/readme"
+[SCS]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:45]:
```suggestion
    rev = "ef7cb4e38e987fcbbdeebd50bd4c211633328150"
```

File: docs/solvers.toml [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+[SDPA]
+    rev = "5d15d3c292cd25b461f524aaaf8eb5ecb8e08b0d"
+[SDPNAL]
+    rev = "87b516fd3db316de8aa12681295138df335feed4"
+[SDPT3]
+    rev = "od/readme"
+[SeDuMi]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:53]:
```suggestion
    rev = "47b4497cab3ce59ec7f65cff455bb2014981d4b4"
```

File: docs/solvers.toml [hunk 32]
```diff
@@ -0,0 +1,103 @@
...
+[MultiObjectiveAlgorithms]
+    rev = "b4413a1ef3b7d28045cedba4e68e5e4959c8bb12"
+    has_html = true
+[Pajarito]
+    rev = "od/readme"
+[Pavito]
+    rev = "od/readme"
+[Penopt]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:43]:
```suggestion
    rev = "486f07d3f3a11f12012ea3ada702a3ee55c8fdc5"
```

File: docs/solvers.toml [hunk 33]
```diff
@@ -0,0 +1,103 @@
...
+[MosekTools]
+    rev = "od/readme"
+[MultiObjectiveAlgorithms]
+    rev = "b4413a1ef3b7d28045cedba4e68e5e4959c8bb12"
+    has_html = true
+[Pajarito]
+    rev = "od/readme"
+[Pavito]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:41]:
```suggestion
    rev = "c0f3d81d9e9fff2963b872cf84fea35bfcf8b1d1"
```

File: docs/solvers.toml [hunk 34]
```diff
@@ -0,0 +1,103 @@
...
+[DiffOpt]
+    rev = "od/readme"
+[Dualization]
+    rev = "od/readme"
+[ECOS]
+    rev = "80d1a927b05b04360b7a8e6800f5ae13bc6a1494"
+[GLPK]
+    rev = "dd835dc425fe8995d965da8a57b4a3a1764cf8da"
+[Gurobi]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:24]:
```suggestion
    rev = "72ee279373d7559e32050c35735d3fc803a7bb12"
```

File: docs/solvers.toml [hunk 35]
```diff
@@ -0,0 +1,103 @@
...
+[Gurobi]
+    rev = "72ee279373d7559e32050c35735d3fc803a7bb12"
+[HiGHS]
+    rev = "b0ab8361558fc63873a5cf380d335409f02d8863"
+[Ipopt]
+    rev = "od/readme"
+[KNITRO]
+    rev = "3fb7c61e886b1b06d9f5cd646e194da638c93355"
+[MiniZinc]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:32]:
```suggestion
    rev = "21d5829c2838959e2ec2d2952495c83f43b271b7"
```

File: docs/solvers.toml [hunk 36]
```diff
@@ -0,0 +1,103 @@
...
+[MosekTools]
+    rev = "od/readme"
+[MultiObjectiveAlgorithms]
+    rev = "b4413a1ef3b7d28045cedba4e68e5e4959c8bb12"
+    has_html = true
+[Pajarito]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:39]:
```suggestion
    rev = "2c008ee52bff34c41ebd394e2daae9eb32387e4b"
```

File: docs/solvers.toml [hunk 37]
```diff
@@ -0,0 +1,103 @@
...
+[HiGHS]
+    rev = "b0ab8361558fc63873a5cf380d335409f02d8863"
+[Ipopt]
+    rev = "od/readme"
+[KNITRO]
+    rev = "3fb7c61e886b1b06d9f5cd646e194da638c93355"
+[MiniZinc]
+    rev = "od/readme"
+[MosekTools]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:34]:
```suggestion
    rev = "e2ed9bc174c6aaa68e457077529d49be52ecbc37"
```

File: docs/solvers.toml [hunk 38]
```diff
@@ -0,0 +1,103 @@
...
+[Clp]
+    rev = "3358ee6"
+[CPLEX]
+    rev = "8e6f211"
+[CSDP]
+    rev = "3a2e4c7"
+[DiffOpt]
+    rev = "od/readme"
+[Dualization]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:18]:
```suggestion
    rev = "13974285d5ed6f678e4b96aafb764dad5238aac0"
```

File: docs/solvers.toml [hunk 39]
```diff
@@ -0,0 +1,103 @@
...
+[Cbc]
+    rev = "a78e129e32f11b5c00ee8acea75c3b2204b8297f"
+[Clp]
+    rev = "3358ee69d9918ab011d5a9aa6201ec50d1bad2a4"
+[CPLEX]
+    rev = "8e6f2112e9cc433085ed77de6712f715edead8fc"
+[CSDP]
+    rev = "3a2e4c7c48af7660a5ba258a4cb6549ab0970366"
+[DiffOpt]
+    rev = "od/readme"
```
Comment (odow) [inline docs/solvers.toml:16]:
```suggestion
#     rev = "od/readme"
```

File: docs/solvers.toml [hunk 40]
```diff
@@ -0,0 +1,103 @@
...
+    rev = "f5f4c9c8f27e0ef8644d65fb956f3bca5305b45e"
+[Cbc]
+    rev = "a78e129e32f11b5c00ee8acea75c3b2204b8297f"
+[Clp]
+    rev = "3358ee69d9918ab011d5a9aa6201ec50d1bad2a4"
+[CPLEX]
+    rev = "8e6f2112e9cc433085ed77de6712f715edead8fc"
+[CSDP]
+    rev = "3a2e4c7c48af7660a5ba258a4cb6549ab0970366"
+[DiffOpt]
```
Comment (odow) [inline docs/solvers.toml:15]:
```suggestion
# [DiffOpt]
```

File: docs/solvers.toml [hunk 41]
```diff
@@ -0,0 +1,103 @@
...
+[SDPNAL]
+    rev = "87b516fd3db316de8aa12681295138df335feed4"
+[SDPT3]
+    rev = "02f69fb87941691722176f77a48933ff7f7a78b4"
+[SeDuMi]
+    rev = "47b4497cab3ce59ec7f65cff455bb2014981d4b4"
+[Xpress]
+    rev = "b98347c49ba05953e29e5a1ef27d0d161780e19d"
+
+# ======================= Updated non-jump-dev packages ====================== #
```
Comment (odow) [inline docs/solvers.toml:57]:
```suggestion
# =========================== non-jump-dev packages ========================== #
```

**Comment by odow** [comment]
Phew. Nearly there. I've omitted DiffOpt for now: https://github.com/jump-dev/DiffOpt.jl/pull/236

You don't really realize how many different solvers we support until you try simultaneous PRs to ~30 packages...

**Comment by blegat** [comment]
I really like this. Any reason why some solvers don't have the MOI API section yet ? I've spotted SDPNAL and SDPT3

**Comment by matbesancon** [comment]
what would be the steps for a solver to be conform to what is needed? I can do SCIP and re-add it

**Comment by odow** [comment]
> what would be the steps for a solver to be conform to what is needed? I can do SCIP and re-add it

It's up to you really, but the more uniform with the others the better. 

So make it look like the other READMEs. Add sections for Affiliation, License, Installation, Use with JuMP, Options. No typos, etc.

File: docs/solvers.toml [hunk 42]
```diff
@@ -0,0 +1,103 @@
...
+[Penopt]
+    rev = "486f07d3f3a11f12012ea3ada702a3ee55c8fdc5"
+[SCS]
+    rev = "ef7cb4e38e987fcbbdeebd50bd4c211633328150"
+[SDPA]
+    rev = "5d15d3c292cd25b461f524aaaf8eb5ecb8e08b0d"
+[SDPNAL]
+    rev = "87b516fd3db316de8aa12681295138df335feed4"
+[SDPT3]
+    rev = "02f69fb87941691722176f77a48933ff7f7a78b4"
```
Comment (odow) [inline docs/solvers.toml:51]:
```suggestion
    rev = "b565aac2a58818090d521f2340e71f597688e4fb"
```

File: docs/solvers.toml [hunk 43]
```diff
@@ -0,0 +1,103 @@
...
+[Pavito]
+    rev = "c0f3d81d9e9fff2963b872cf84fea35bfcf8b1d1"
+[Penopt]
+    rev = "486f07d3f3a11f12012ea3ada702a3ee55c8fdc5"
+[SCS]
+    rev = "ef7cb4e38e987fcbbdeebd50bd4c211633328150"
+[SDPA]
+    rev = "5d15d3c292cd25b461f524aaaf8eb5ecb8e08b0d"
+[SDPNAL]
+    rev = "87b516fd3db316de8aa12681295138df335feed4"
```
Comment (odow) [inline docs/solvers.toml:49]:
```suggestion
    rev = "00a3fa19f4e1235587948113b0b681da17f4dab5"
```

**Comment by odow** [comment]
I talked to @blegat last night about this. There is still plenty of changes we could make to the READMEs, but it's been a bit of a slog to get to this point, so my preference is to merge this, and then deal with the changes one-by-one, rather than playing this juggling game of updating 26 different READMEs at the same time.

**Comment by chriscoey** [comment]
Thanks for all the effort here! This will be nice

**Comment by odow** [comment]
I wonder if we should have a "Solvers" section, and an "Extensions" section so that we can advertise DiffOpt, SumOfSquares, PolyJuMP, PowerModels, SDDP.jl, InfiniteOpt, etc. There's no reason they couldn't also provide a `JuMP.md` file. It makes the ecosystem a lot more cohesive and easier for new users to find.

**Comment by odow** [comment]
Pyomo has https://pyomo.readthedocs.io/en/stable/contributed_packages/index.html.

**Comment by pulsipher** [comment]
> I wonder if we should have a "Solvers" section, and an "Extensions" section so that we can advertise DiffOpt, SumOfSquares, PolyJuMP, PowerModels, SDDP.jl, InfiniteOpt, etc. There's no reason they couldn't also provide a `JuMP.md` file. It makes the ecosystem a lot more cohesive and easier for new users to find.

I am in favor of this.

File: docs/make.jl [hunk 44]
```diff
@@ -410,9 +465,11 @@ if _PDF
     for (root, dir, files) in walkdir(joinpath(@__DIR__, "src", "tutorials"))
         _remove_literate_footer.(joinpath.(root, dir))
     end
-    # Remove release notes from PDF
-    splice!(_PAGES, 7)   # JuMP release notes
-    pop!(_PAGES[end][2]) # MOI release notes
+    moi = pop!(PAGES)   # remove /MathOptInterface
```
Comment (odow) [inline docs/make.jl:468]:
```suggestion
    moi = pop!(_PAGES)   # remove /MathOptInterface
```

**Comment by odow** [comment]
Okay, this is in a reasonable state. It now has a section for solvers and extensions. I'm in favor of merging this and adding new pages in a separate PR. No need to make this PR long-lived.

File: docs/packages.toml [hunk 45]
```diff
@@ -0,0 +1,114 @@
...
+[SDPA]
+    rev = "5d15d3c292cd25b461f524aaaf8eb5ecb8e08b0d"
+[SDPNAL]
+    rev = "00a3fa19f4e1235587948113b0b681da17f4dab5"
+[SDPT3]
+    rev = "b565aac2a58818090d521f2340e71f597688e4fb"
+[SeDuMi]
+    rev = "47b4497cab3ce59ec7f65cff455bb2014981d4b4"
+[SumOfSquares]
+    rev = "od/readme"
```
Comment (odow) [inline docs/packages.toml:59]: TODO

**Review by blegat — APPROVED** [review]

**Pull request merged**